documentation/config_file_auth.md

```diff
@@ -9,142 +9,168 @@ canonical: "/puppetserver/latest/config_file_auth.html"
 
-In a HOCON rule, use the `query-params` setting and provide the list of query parameters as an array to the setting's `environment` parameter:
+For example, this rule would match a request URL containing the `environment=production` or `environment=test` query parameters:
 
-~~~
+~~~ hocon
 query-params: {
     environment: [ production, test ]
 }
 ~~~
 
-In a deprecated `auth.conf` rule, the `environment` parameter adds a comma-separated list of query parameters as a suffix to the base URL.
+> **Note:** For Puppet 3 master endpoints, the `environment` was represented as the first subpath in the URL instead of as a query parameter. As noted in the [Puppet 3 agent compatibility section](#puppet-3-agent-compatibility), Puppet Server translates incoming Puppet 3-style URLs to Puppet 4-style URLs before evaluating them against the new HOCON `auth.conf` rules, so the `query-params` approach above replaces environment-specific rules for both Puppet 3 and Puppet 4.
 
-~~~
-environment: production,test
-~~~
+#### `allow`, `allow-unauthenticated`, and `deny`
 
-> **Note:** For Puppet 3 master endpoints, the `environment` was represented as the first subpath in the URL instead of as a query parameter. As noted in the [Puppet 3 agent compatibility section](#puppet-3-agent-compatibility), Puppet Server translates incoming Puppet 3-style URLs to Puppet 4-style URLs before evaluating them against the new HOCON `auth.conf` rules, so the `query-params` approach above replaces environment-specific rules both for Puppet 3 and Puppet 4.
+After each rule's `match-request` section, it must also have an `allow`, `allow-unauthenticated`, or `deny` parameter. (You can set both `allow` and `deny` parameters for a rule, though Puppet Server always prioritizes `deny` over `allow` when a request matches both.) 
 
-### `allow-unauthenticated` and `auth`
+If a request matches the rule, Puppet Server checks the request's authenticated "name" (see [`allow-header-cert-info`](#allow-header-cert-info)) against these parameters to determine what to do with the request.
 
-In the HOCON rule format, Puppet Server determines whether to allow or deny requests _after_ a rule is matched. By default, this authentication method refuses unauthenticated requests. You can force Puppet Server to allow requests from unauthenticated clients by setting the Boolean `allow-unauthenticated` parameter to true:
+* **`allow-unauthenticated`**: If this Boolean parameter is set to `true`, Puppet Server allows the request --- even if it can't determine an authenticated name. **This is an insecure configuration and not recommended for production environments.** A rule with this parameter set to `true` can't also contain the `allow` or `deny` parameters.
```

Comment:
We may want to soften or eliminate the language about "... not recommended for production environments" for the `allow-unauthenticated: true` setting.  There are some situations in our own rule defaults where we have to do this.  The endpoints for submitting a certificate request and retrieving back the signed certificate, for example, use `allow-unauthenticated: true` because the submitter has to hit those without a certificate first in order to get the certificate that it will use for other requests later on.